Update dependencies, workflows, and configurations (mostly client side)

kklamberty opened the pull request:
The focus of this PR is to get the structure and requirements from client side of the Iteration Template into this lab. This is likely to break things. It is unlikely to pass the checks until at least the changes within the Angular files are up to date. I did change the github actions for both client and server here because I was not thinking.

kklamberty committed:
Update `client.yaml`

Use newer versions of actions and change to Node v18. This also included adding some Cypress environment information that has been added to the iteration template.

kklamberty committed:
Update `e2e.yaml`

Use newer versions of several GitHub actions. Update to Node v18

kklamberty committed:
Update `server.yml`

Use updated versions of GitHub actions

kklamberty committed:
Update `settings.yml` to remove Better Code Hub

Since it's not available anymore, remove BCH as a required check. There is an issue related to finding a replacement that exists on the Iteration Template.

kklamberty committed:
Include Thunder Client in recommended extensions

Update extensions.json to include Thunder Client in the list of recommended extensions for VS Code.

kklamberty committed:
Remove code quality checks

Since the code quality checks we were using (BCH, LGTM) are not available, remove them and references to them.

kklamberty committed:
Update Cypress configuration and file structure

This includes replacing the configuration json file with a typescript file and moving the e2e tests to a folder called e2e. I believe that there was a time when Cypress was using a folder called `integration` by default (maybe to differentiate it from Protractor tests?), but now it seems to be e2e by default, so we will follow suit. 

Also, there is some problem we've had with Chai assertions and Jasmine assertions not playing nicely together and causing lots of errors to show up in VS Code. The change to tsconfig.json makes that problem go away.

kklamberty committed:
Make `tsconfig.json` match the Iteration Template

Minor changes that make this config match the Iteration Template. One of these changes may be somewhat major, I guess. This updates to ES2022. I noticed that 2023 was an option, and I don't know if that is something to look into.

kklamberty committed:
Rename some Cypress files

This has no impact on the contents of the files... literally just the names.

kklamberty committed:
Update dependencies in `package.json`

This upgrades a bunch of things to newer versions. I'm leaving out a couple to do separately so that they are seen as changes and not as replacements because I think it's more informative if it shows differences in more detail. There are not many new things---just updated.

kklamberty committed:
Alphabetize modules and imports in `app.module.ts`

It was very difficult to find differences in this file without knowing a little bit about where to look. I think this reorganization makes it easier.

kklamberty committed:
Remove output folder

We will just use the default. Something was different and broken in Cypress, and I was trying to figure out what. It's not clear that this was causing any problem, but it was different from what we have in the Iteration Template, so I'm taking it out for consistency.

kklamberty committed:
Add dependency from Iteration Template

The one thing I need to add (aside from the docker stuff that we don't use here) from the Iteration Template to make this match what we have there. For some reason, putting this in for the previous update made the changes look much more drastic for the other commit about dependencies, so I held onto it for its own step.

kklamberty committed:
Update `package-lock.json`

I suspect this will make a lot of dependabot things go away, but maybe not.

kklamberty committed:
Match Iteration Template

kklamberty committed:
Add part of the upgrade to Angular v15

This part of the style relates specifically to the upgrade to Angular v15

kklamberty committed:
Update to Angular v15

kklamberty committed:
Move away from flex-layout

kklamberty committed:
Remove comment about lowercase 'l'

This comment in the code is not relevant to this code. I believe it is from another test where `maxlength` was used since that is commented on again in the code at that location.

kklamberty committed:
Move away from flex-layout

kklamberty committed:
Change to `UntypedFormGroup`

This change matches our Iteration Template and might be important since `FormGroup` doesn't mean what it used to now that `FormGroup` elements are strongly typed by default. Since I'm not updating using the CLI, this kind of thing could be missed easily. It might not be a problem, but just in case it is, I will make it match what would have happened with the CLI update.

See https://github.com/UMM-CSci-3601/3601-iteration-template/issues/977 for more information. In particular, I'm not sure if our `FormGroup` will work because I think it might need to have the same type for all of the elements, and ours does not. It's possible that means within the options for a single input. I'd have to read more carefully.

kklamberty committed:
Change appearance

We added this when we moved away from flex-layout because things were all kind of mashed together. The outline helped differentiate one card from another. It's probably not a big important change.

kklamberty committed:
Move away from flex-layout

kklamberty committed:
Update `user-list.component.spec.ts` to match Iteration Template

After bringing some of the documentation from here into the Iteration Template, bring the test back here to Lab 4 because there are other updates to modernize the test.

kklamberty committed:
Update `user-list.component.ts` to match Iteration Template

This also happens to make the usage of subscribe follow the not deprecated format.

kklamberty committed:
Update `polyfills.ts` to match Iteration Template

kklamberty committed:
Update `angular.json` to match Iteration Template

kklamberty committed:
Move away from flex-layout

kklamberty committed:
Update `test.ts` to match Iteration Template

kklamberty committed:
Update `user.service.spec.ts`(to match a branch in Lab 3)

In this case, I actually pulled the Karma tests from the branch in lab 3 where we are working to modernize the Karma tests. There was such nice work there that I could not let it go to waste. I also added a test for adding users. So, this doesn't match the Iteration Template at this point.

kklamberty committed:
Change activatedRoute to use the constructor

Using the constructor for ActivatedRouteStub lets us improve our coverage. We alter the stub using setParamMap in the tests as needed. This doesn't break anything and it does let us exercise the use of the constructor, which seems like a fine plan.

kklamberty committed:
Update comment to reflect a need for further exploration

I think it might be good to look at this further, but it at least helps us know a little about adding users.

kklamberty committed:
Update e2e tests

Several updates were needed in the e2e tests after upgrading to Angular v15.

kklamberty committed:
Make GitHub happy that I noticed a conflict in `package-lock.json`

This is me recognizing that there is a merge conflict and not really handling it because I'll regenerate it and replace it anyhow.

kklamberty commented on client/src/app/app.module.ts:
```diff
@@ -21,56 +23,52 @@ import { MatSnackBarModule } from '@angular/material/snack-bar';
+  MatTooltipModule,
 ];
 
 @NgModule({
```
I also tried to alphabetize elsewhere. @floogulinc suggests (correctly) that order can matter for these, so I'm going to make another commit that puts back the dangerous things into the order they were in before. This didn't seem to cause problems, but it was an arbitrary decision just meant to help readability. If it's not safe to do, better to leave it confusing.

kklamberty committed:
Undo some alphabetization

I also tried to alphabetize elsewhere. @floogulinc suggests (correctly) that order can matter for these, so I'm going to make another commit that puts back the dangerous things into the order they were in before. This didn't seem to cause problems, but it was an arbitrary decision just meant to help readability. If it's not safe to do, better to leave it confusing.

https://github.com/UMM-CSci-3601/3601-iteration-template/pull/1293#discussion_r1103957374

kklamberty commented:
My change to the `package-lock.json` file seems to have taken out a lot of "resolved" and "integrity" elements, and I'm not sure if that matters. See: [5c90459](https://github.com/UMM-CSci-3601/3601-lab4-mongo/pull/55/commits/5c90459cbaac9682b607e3230175a62d65ab7d0c) If you would like to verify any of that.

NicMcPhee committed:
Update the last few `npm` packages

There were three packages that still weren't up-to-date according to `npm outdated` and this updates those:

- `eslint-plugin-import`
- `eslint-plugin-jsdoc`
- `zone.js`

NicMcPhee commented on .github/workflows/client.yaml:
```diff
@@ -12,23 +12,25 @@ jobs:
+      uses: actions/cache@v3
       with:
         path: ~/.npm
         key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
```
You still have all the old (and I think non-functional) caching, but aren't using the new `npm` caching command. I would definitely fix that as I think it will speed things up for the students.

kklamberty replied:
Can you help -- is this a comment on line 21? This matches what we have in the iteration database, and I thought we were using the newest one there.

NicMcPhee commented on .github/workflows/e2e.yaml:
```diff
@@ -14,18 +14,18 @@ jobs:
+      uses: actions/cache@v3
       with:
         path: ~/.npm
         key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
```
Same issue with the caching here as well.

NicMcPhee commented on client/src/app/users/add-user.component.ts:
```diff
@@ -12,7 +12,7 @@ import { UserService } from './user.service';
 export class AddUserComponent implements OnInit {
 
-  addUserForm: FormGroup;
+  addUserForm: UntypedFormGroup;
```
Thinking about it, we almost certainly want to move to typed form groups since then the type checker can help us out more. We should probably create an issue for that.

kklamberty replied:
I agree; It just happened to not seem broken when it was just listed as being a FormGroup... but I'm not sure if that's meaningful TBH. So, I went ahead and made this a noticeable thing to address.

NicMcPhee reviewed (changes requested):
This looks great, although I do think you want to fix the `npm` caching business in two of the `.github/workflows` files.

Thanks for grinding through this huge amount of work. We really do need to find a better way to manage these complex collections of repositories so we don't repeat this work all the time.

kklamberty committed:
Use up-to-date node caching in `client.yaml`

Note: this does not cache node_modules (as per: https://github.com/actions/setup-node#caching-global-packages-data)

jpwalbran commented:
This looks like it does what we want it to:
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows
Haven't dug into it yet.

kklamberty committed:
Add new caching

NicMcPhee committed:
Merge branch 'main' into update-dependencies-workflows-and-configs

kklamberty committed:
Update cache-dependency-path

kklamberty committed:
Merge remote-tracking branch 'origin/update-dependencies-workflows-and-configs' into update-dependencies-workflows-and-configs

kklamberty committed:
Add dashes to fix what goes together

kklamberty committed:
Try again

NicMcPhee committed:
Fix broken workflow file

NicMcPhee committed:
Merge remote-tracking branch 'refs/remotes/origin/update-dependencies-workflows-and-configs'

Conflicts:
	.github/workflows/client.yaml

NicMcPhee committed:
Fix indentation issue in client workflow

Hopefully this fixes the GitHub Actions workflow problems.

NicMcPhee committed:
Fix more indentation issues in client workflow file

Hopefully (again) this fixes the workflow build problems in GitHub Actions.

floogulinc commented on .github/workflows/e2e.yaml:
```diff
@@ -14,18 +14,18 @@ jobs:
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
 
     - name: Cache node modules
```
Now that you're having the caching from `actions/setup-node` you can remove this entire step.

floogulinc replied:
@kklamberty did you merge without removing this?

kklamberty replied:
Yes. I'm not confident that you are right (or wrong) about this step being unneceassary. The node cache step specifically says it does not cache node_modules, and this step aims to do that. We are not sure if it's working or not working, but it seems that it's _different_ in the least. 

This step does not cache node_modules, as is noted here: https://github.com/actions/setup-node
Different sources have varying opinions on whether or not we *should* cache node_modules, but since our step name suggests that we were trying to do so, I didn't think it would be problematic to keep our attempt in here.

In fact - @NicMcPhee and @jpwalbran - I think that looking at what to do if we have a cache hit might do the trick.
https://github.com/actions/cache#skipping-steps-based-on-cache-hit - I don't think we were ever looking for the cache and behaving differently based on whether or not we had a cache hit.

floogulinc replied:
This step _isn't_ caching the `node_modules` directory (which we shouldn't be). It sets up caching just for `~/.npm` (see the `path` below) which is the same as what setup-node is doing.

floogulinc replied:
`~/.npm` is a directory npm uses specifically for caching. We should not cache `node_modules` as that wouldn't work with `npm ci` and could break when the node version changes between runs.

floogulinc commented on client/src/app/users/user-profile.component.html:
```diff
@@ -1,5 +1,5 @@
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
```
Looks like you've left some stuff from Angular flex-layout here. Hasn't that been removed?

floogulinc commented on client/src/app/users/user-list.component.html:
```diff
@@ -59,25 +59,25 @@
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10" *ngIf="serverFilteredUsers; else usersError" >
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10" *ngIf="serverFilteredUsers; else usersError" >
```
Same here there are attributes for the old flex layout library that don't do anything still here and probably in many other places.

floogulinc commented on client/src/app/users/user-list.component.html:
```diff
@@ -1,11 +1,11 @@
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
```
See previous 2 comments.

floogulinc commented:
I'm not going to leave individual comments for each of them but it seems in many places there are still attributes in html elements for the Angular flex-layout library that was removed. Anything starting with `fxFlex` for example won't be doing anything anymore. This was probably missed in the iteration template as well.
